File: Leetcode/977. Squares of a Sorted Array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {      
        int a=-1, b=-1, n = nums.size();
        vector<int> ans;
        
        for(int i=0; i<n; i++) {
            if(nums[i] >= 0) {
                b = i;
                a = i-1;
                break;
            }
        }
        
        if(b == -1) {
            b = n;
            a = n-1;
        }
        
        while(a >=0 || b < n) {
            if(a < 0) ans.push_back(pow(nums[b++],2));
            else if(b >= n) ans.push_back(pow(nums[a--],2));
            else if(abs(nums[a]) < nums[b]) ans.push_back(pow(nums[a--],2));
            else ans.push_back(pow(nums[b++],2));
        }
        
        return ans;
    }
};

File: Leetcode/Rust/977. Squares of a Sorted Array.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut ans = vec![0; n];
        let mut idx = n - 1;
        let (mut l, mut r) = (0, n - 1);
        
        while l < r {
            println!("l = {l}; r = {r}");
            let a = nums[l] * nums[l];
            let b = nums[r] * nums[r];
            if a > b {
                ans[idx] = a;
                l += 1;
            } else {
                ans[idx] = b;
                r -= 1;
            }
            idx -= 1;
        }
        
        ans[idx] = nums[l] * nums[l];
        
        ans
    }
}
